Optimize mobile model on cloned module instead of in-place transformation

xcheng16 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36357 Python binding api to optimize for mobile model on script module.
* **#36621 Optimize mobile model on cloned module instead of in-place transformation**

Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

Commit:
Optimize mobile model on cloned module instead of in-place transformation

Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -297,11 +297,13 @@ void FoldPrePackingOps(script::Module& m) {
-  insertPrePackedOps(m);
-  m = freeze_module(m);
-  FoldPrePackingOps(m);
+  auto moduleClone = module.clone();
```
Do variable names follow camel case? From what I noticed no. If so change accordingly. Otherwise look great.

xcheng16 replied:
good point, seems like not.

kimishpatel reviewed (changes requested):
Small change.

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

kimishpatel reviewed (approved):
LGTM. There are merge conflict. Plus import to phabricator.

AshkanAliabadi commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -297,11 +297,13 @@ void FoldPrePackingOps(script::Module& m) {
+  insertPrePackedOps(cloned_module);
+  cloned_module = freeze_module(cloned_module);
+  FoldPrePackingOps(cloned_module);
+  return cloned_module;
```
This function's return type is void, right?  You are probably getting a compiler warning which PyTorch's compilation settings do not treat as error.  Also if we are no longer modifying the parameter please pass it by const reference.

AshkanAliabadi replied:
It's still safer to pass the argument by const ref.  Please consider that if there are no other complications.  Otherwise we wouldn't have a way of knowing if we accidentally modify m.

AshkanAliabadi reviewed (changes requested):
When it comes making modifications in place vs not, what I have seen Python quantization modules do is to take an inplace argument, which if set to true, modifies the model in place, otherwise creates a copy.  This is merely a suggestion, but following the principle of least surprise and API consistency I think we could follow the same principle.  Again, just a suggestion, this is also fine as well.

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

kimishpatel commented:
> When it comes making modifications in place vs not, what I have seen Python quantization modules do is to take an inplace argument, which if set to true, modifies the model in place, otherwise creates a copy. This is merely a suggestion, but following the principle of least surprise and API consistency I think we could follow the same principle. Again, just a suggestion, this is also fine as well.

I agree, but I think common expectation is to not do inplace? So if inplace is wanted, you have to be explicit about it. Also, something funny was happening when it was doing inplace mutation silently. For some reason it wasn't working until the module was serialized and then deserialized. That needs to be looked at, but this change at least make it cleaner and explicit.

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

Commit:
Update on "Optimize mobile model on cloned module instead of in-place transformation"


Summary:
Instead of doing in-place transformation inside optimizeForMobile method,
we would like to maintain the original structure for passed scriptedModule,
so before optmization starts, we use the cloned module to do subsequent optimization
process and return the optimized cloned module.

Test Plan:
unit test
python test/test_mobile_optimizer.py
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21028406](https://our.internmc.facebook.com/intern/diff/D21028406)

[ghstack-poisoned]

AshkanAliabadi reviewed (approved):
Thank you!
